adapter/serviceregistry/aggregate/controller.go

```diff
@@ -0,0 +1,125 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package aggregate
+
+import (
+	"github.com/golang/glog"
+
+	"istio.io/pilot/model"
+	"istio.io/pilot/platform"
+)
+
+// Controller aggregates data across different registries and monitors for changes
+type Controller struct {
+	registries map[platform.ServiceRegistry]model.Controller
+}
+
+// NewController creates a new Aggregate controller
+func NewController(registries map[platform.ServiceRegistry]model.Controller) *Controller {
+	return &Controller{
+		registries: registries,
+	}
+}
+
+// Services list declarations of all services in the system
+func (c *Controller) Services() []*model.Service {
+	services := make([]*model.Service, 0)
+	for _, r := range c.registries {
+		services = append(services, r.(model.ServiceDiscovery).Services()...)
```

Comment:
What about conflicting services? For example, if both Eureka and Consul have a service with the same hostname?

Same question for the rest of these methods.

Reply:
No support for conflicting services. It's user responsibility to maintain uniqueness.